pkg/extractor/extractor.go

```diff
@@ -0,0 +1,121 @@
+package extractor
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/mt-sre/addon-metadata-operator/pkg/extractor/bundle"
+	"github.com/mt-sre/addon-metadata-operator/pkg/extractor/index"
+	imageparser "github.com/novln/docker-parser"
+	"github.com/operator-framework/operator-registry/pkg/registry"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+type mainExtractor struct {
+	log    *logrus.Logger
+	index  index.Extractor
+	bundle bundle.Extractor
+}
+
```

Comment:
This file (or at least this package) is where the `IndexExtractor` and `BundleExtractor` interfaces should be defined since this is the package where they are needed. The `extractor/index` and `extractor/bundle` packages only implement an instance of these interfaces.

Reply:
I see the `extractor/index` and `extractor/bundle`, as their own thing, they could be use without the main extractor. And defining all the interfaces in `pkg/extractor/interfaces.go` would create a cyclic import graph.

I liked the extra nesting, similar to `net/http`, where the namespace clearly express what these subpackages are about. Also solves the naming collision for `Extractor`, `Cache`, `Encoder`, `WithLog`, `ExtractorOption`, etc. and allows us to have simple naming and avoid prepending/prefixing everything with `Bundle` or `Index`, including filenames.

Reply:
I see. Interesting. I can flatten the structure.